File: src/features/typing/data/java/partition_labels.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    public List<Integer> partitionLabels(String s) {
        Map<Character, Integer> last = new HashMap<>();
        for (int i = 0; i < s.length(); i++) {
            last.put(s.charAt(i), i);
        }
        List<Integer> result = new ArrayList<>();
        int start = 0;
        int end = 0;
        for (int i = 0; i < s.length(); i++) {
            if (last.get(s.charAt(i)) > end) {
                end = last.get(s.charAt(i));
            }
            if (i == end) {
                result.add(i - start + 1);
                start = i + 1;
            }
        }
        return result;
    }
}

File: src/features/typing/data/ts/partition_labels.ts
function partitionLabels(s: string): number[] {
    const last = new Map<string, number>();
    for (let i = 0; i < s.length; i++) {
        last.set(s[i], i);
    }
    const result: number[] = [];
    let start = 0;
    let end = 0;
    for (let i = 0; i < s.length; i++) {
        if (last.get(s[i])! > end) {
            end = last.get(s[i])!;
        }
        if (i === end) {
            result.push(i - start + 1);
            start = i + 1;
        }
    }
    return result;
}
